File: leetcode/company-wise/netflix/0460_lfu_cache.py
# LC 460. LFU Cache | Hard | Netflix
from collections import defaultdict, OrderedDict

class LFUCache:
    def __init__(self, capacity):
        self.cap = capacity; self.min_freq = 0
        self.key_val = {}; self.key_freq = {}
        self.freq_keys = defaultdict(OrderedDict)

    def get(self, key):
        if key not in self.key_val: return -1
        self._update(key); return self.key_val[key]

    def put(self, key, value):
        if self.cap <= 0: return
        if key in self.key_val: self.key_val[key] = value; self._update(key); return
        if len(self.key_val) >= self.cap:
            evict, _ = self.freq_keys[self.min_freq].popitem(last=False)
            del self.key_val[evict]; del self.key_freq[evict]
        self.key_val[key] = value; self.key_freq[key] = 1
        self.freq_keys[1][key] = None; self.min_freq = 1

    def _update(self, key):
        freq = self.key_freq[key]
        del self.freq_keys[freq][key]
        if not self.freq_keys[freq] and freq == self.min_freq: self.min_freq += 1
        self.key_freq[key] = freq + 1
        self.freq_keys[freq + 1][key] = None

File: leetcode/company-wise/netflix/0460_lfu_cache.cpp
// LC 460. LFU Cache | Hard | Netflix
#include <unordered_map>
#include <list>
using namespace std;

class LFUCache {
    int cap, minFreq;
    unordered_map<int, pair<int,int>> kv; // key -> {val, freq}
    unordered_map<int, list<int>> freq_list;
    unordered_map<int, list<int>::iterator> pos;

    void update(int key) {
        int f = kv[key].second;
        freq_list[f].erase(pos[key]);
        if (freq_list[f].empty() && f == minFreq) minFreq++;
        kv[key].second++;
        freq_list[f+1].push_back(key);
        pos[key] = prev(freq_list[f+1].end());
    }
public:
    LFUCache(int capacity) : cap(capacity), minFreq(0) {}
    int get(int key) { if (!kv.count(key)) return -1; update(key); return kv[key].first; }
    void put(int key, int value) {
        if (cap <= 0) return;
        if (kv.count(key)) { kv[key].first = value; update(key); return; }
        if ((int)kv.size() >= cap) {
            int evict = freq_list[minFreq].front();
            freq_list[minFreq].pop_front(); pos.erase(evict); kv.erase(evict);
        }
        kv[key] = {value, 1}; freq_list[1].push_back(key);
        pos[key] = prev(freq_list[1].end()); minFreq = 1;
    }
};
